joerd/download.py

```diff
@@ -0,0 +1,169 @@
+      * 'verifier' - A function which is called with a filelike object. It
+        should return True if the file is okay and appears to be fully
+        downloaded.
+    """
+    logger = logging.getLogger('download')
+
+    with closing(tempfile.NamedTemporaryFile()) as tmp:
+        # current file position = number of bytes read
+        filepos = 0
+
+        # file size when downloaded, if known
+        filesize = None
+
+        # number of attempts so far
+        tries = 0
+
+        # maximum number of attempts to make
+        max_tries = options.get('tries', 1)
+
+        # timeout for blocking operations (e.g: connect) in seconds
+        timeout = options.get('timeout', 60)
+
+        # verifier function
+        verifier = options.get('verifier')
+
+        # backoff function - to delay between retries
+        backoff = options.get('backoff')
+
+        # whether the server supports Range headers (if it doesn't we'll have
+        # to restart from the beginning every time).
+        accept_range = False
+
+        # we need to download _something_ if the file position is less than the
+        # known size, or the size is unknown.
+        while filesize is None or filepos < filesize:
+            # explode if we've exceeded the number of allowed attempts
+            if tries >= max_tries:
+                raise DownloadFailedError("Max tries exceeded (%d) while "
+                                          "downloading file %r"
+                                          % (max_tries, url))
+            else:
+                if backoff:
+                    backoff(tries)
+                tries += 1
+
+            req = urllib2.Request(url)
+
+            # if the server supports accept range, and we have a partial
+            # download then attemp to resume it.
+            if accept_range and filepos > 0:
+                logger.info("Continuing (try %d/%d) at %d bytes: %r"
+                            % (tries, max_tries, filepos, url))
+                assert filesize is not None
+                req.headers['Range'] = 'bytes=%s-%s' % (filepos, filesize - 1)
+            else:
+                # otherwise, truncate the file in readiness to download from
+                # scratch.
+                logger.info("Downloading (try %d/%d) %r"
+                            % (tries, max_tries, url))
+                filepos = 0
+                tmp.seek(0, os.SEEK_SET)
+                tmp.truncate(0)
+
+            try:
+                f = urllib2.urlopen(req, timeout=timeout)
+
+            except (IOError, httplib.HTTPException) as e:
+                logger.debug("Got HTTP error: %s" % str(e))
+                continue
+
+            except ftplib.all_errors as e:
+                logger.debug("Got FTP error: %s" % str(e))
+                continue
+
+            # try to get the filesize, if the server reports it.
+            if filesize is None:
+                filesize = int(f.info().get('Content-Length'))
```

Comment:
Do we have to be more careful here if `Content-Length` doesn't exist? Otherwise, will it call `int(None)` and blow up?

```python
content_length = f.info().get('Content-Length')
if content_length is not None:
    try:
        filesize = int(content_length)
    except ValueError:
        pass
```

The check for an integer might be a little pedantic though :)